source/common/common/perf_annotation.h

```diff
@@ -0,0 +1,162 @@
+ * can then be reported multiple times.
+ */
+#define PERF_OPERATION(perf_var) Envoy::PerfOperation(perf_var)
+
+/**
+ * Records performance data initiated with PERF_OPERATION. The category and description
+ * are joined with in the library, but only if perf is enabled. This way, any concatenation
+ * overhead is skipped when perf-annotation is disabled.
+ */
+#define PERF_RECORD(perf, category, description)                                                   \
+  do {                                                                                             \
+    perf.record(category, description);                                                            \
+  } while (false)
+
+/**
+ * Dumps recorded performance data to stdout. Expands to nothing if not enabled.
+ */
+#define PERF_DUMP() Envoy::PerfAnnotationContext::dump()
+
+/**
+ * Returns the aggregated performance data as a formatted multi-line string, showing a
+ * formatted table of values. Returns "" if perf-annotation is disabled.
+ */
+#define PERF_TO_STRING() Envoy::PerfAnnotationContext::toString()
+
+/**
+ * Clears all performance data.
+ */
+#define PERF_CLEAR() Envoy::PerfAnnotationContext::clear()
+
+/**
+ * Controls whether performacne collection and reporting is thread safe. For now,
+ * leaving this enabled for predictability across multiiple applications, on the assumption
+ * that an uncontended mutex lock has vanishingly small cost. In the future we may try
+ * to make this system thread-unsafe if mutext contention disturbs the metrics.
+ */
+#define PERF_THREAD_SAFE true
+
+namespace Envoy {
+
+/**
+ * Defines a context for collecting performance data. Note that this class is
+ * fully declared and defined even if ENVOY_PERF_AUTOMATION is off. We depend on
+ * the macros to disable performance collection for production.
+ */
+class PerfAnnotationContext {
+public:
+  /**
+   * Records time consumed by a category and description, which are just
+   * joined together in the library with " / ".
+   */
+  void record(std::chrono::nanoseconds duration, absl::string_view category,
+              absl::string_view description);
+
+  /**
+   * Renders the aggregated statistics as a string.
+   * @return std::string the performance data as a formatted string.
+   */
+  static std::string toString();
+
+  /**
+   * Dumps aggregated statistics (if any) to stdout.
+   */
+  static void dump();
+
+  /**
+   * Thread-safe lazy-initialization of a PerfAnnotationContext on first use.
+   * @return PerfAnnotationContext* the context.
+   */
+  static PerfAnnotationContext* getOrCreate();
+
+  /**
+   * Clears out all aggregated statistics.
+   */
+  static void clear();
+
+private:
+  /**
+   * PerfAnnotationContext construction should be done via getOrCreate().
+   */
+  PerfAnnotationContext();
+
+  typedef std::pair<std::chrono::nanoseconds, uint64_t> DurationCount;
+  typedef std::map<std::string, DurationCount> DurationCountMap;
```

Comment:
Originally, yes, but then I changed the output function to sort by time rather than name.

Also, I also got into a habit from before there was an official unordered_map of using std::set or std::map unless I had a specific performance reason not to, just for consistent iteration order of tests and predictably good performance without outliers.  Your comment suggests that is not the envoy culture so I'll switch.

Reply:
I tried changing to an unordered_map, and ran into a heap of trouble getting it to compile.  Other hash_map implementations I've used before unordered_map would have been very easy; just change the declaration in the header.

There is some discussion about the differences here, which I didn't fully grasp: https://stackoverflow.com/questions/17172080/insert-vs-emplace-vs-operator-in-c-map .

Reply:
I think general C++ best practice is to prefer unordered_map over map unless ordering matters; this I picked up via Google style rather than the Envoy code base. The key in your example is a string, so not sure why that is problematic when switching to a hash map.

Reply:
The key is now a pair of strings but anyway once I defined an explicit hasher, unordered_map worked great.  The compiler error messages when I lacked one were shockingly unhelpful :)